### ISSUE TITLE
Pod status becomes MatchNodeSelector after restart kubelet

### ISSUE THREAD

**Issue by weiwei04** [kubernetes/kubernetes#99708]
#### What happened:

step 1: label node0 with region=r0
step 2: schedule a pod A with labelSelector region=r0 to node0, and pod A up and running
step 3: label node0 with region=rx
step 4: restart kubelet on node0, after kubelet restart, pod A evicted by kubelet and status becomes MatchNodeSelector

#### What you expected to happen:

pod A up and running on node0 even after we change nodeLabel and restart kubelet

#### How to reproduce it (as minimally and precisely as possible):

see `What happened`

#### Environment:
- Kubernetes version (use `kubectl version`): v1.14.4
- Cloud provider or hardware configuration: baremetal
- OS (e.g: `cat /etc/os-release`): -
- Kernel (e.g. `uname -a`): -
- Install tools: -
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by weiwei04** [comment]
Look into kubelet code:

After kubelet restarted, it will list pods from APIServer and go through HandlePodAddition which will call predicates.Admit to check nodeName & labelSelector/Affinity

https://github.com/kubernetes/kubernetes/blob/ce968390b180e466b8b9a8f5a30401b501e69695/pkg/kubelet/kubelet.go#L1884
```golang

func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handler SyncHandler,
	syncCh <-chan time.Time, housekeepingCh <-chan time.Time, plegCh <-chan *pleg.PodLifecycleEvent) bool {
	select {
	case u, open := <-configCh:
		// Update from a config source; dispatch it to the right handler
		// callback.
		if !open {
			klog.Errorf("Update channel is closed. Exiting the sync loop.")
			return false
		}

		switch u.Op {
		case kubetypes.ADD:
			klog.V(2).Infof("SyncLoop (ADD, %q): %q", u.Source, format.Pods(u.Pods))
			// After restarting, kubelet will get all existing pods through
			// ADD as if they are new pods. These pods will then go through the
			// admission process and *may* be rejected. This can be resolved
			// once we have checkpointing.
			handler.HandlePodAdditions(u.Pods)

```

and the running pod A will violate

https://github.com/kubernetes/kubernetes/blob/ce968390b180e466b8b9a8f5a30401b501e69695/pkg/kubelet/lifecycle/predicate.go#L92

**Comment by liggitt** [comment]
is this resolved by https://github.com/kubernetes/kubernetes/pull/94087?

**Comment by weiwei04** [comment]
No, #94087 solves when kubelet restart, list pods before kubelet get node info. In my case, the node label is changed, and the pod running on this node no longer satisfy pod's affinity rule.

step to reproduce:

step 1: label node0 with region=r0
step 2: schedule a pod A with nodeAffinity region=r0 to node0, and pod A up and running
step 3: label node0 with region=rx
step 4: restart kubelet on node0, after kubelet restart, pod A evicted by kubelet and status becomes MatchNodeSelector

**Comment by ehashman** [comment]
>  step 4: restart kubelet on node0, after kubelet restart, pod A evicted by kubelet and status becomes MatchNodeSelector

That sounds like correct/expected behaviour to me though.

> Expected Behavior: pod A up and running on node0 even after we change nodeLabel and restart kubelet

If the workload no longer matches the node selector, why wouldn't it be evicted? It doesn't match the label anymore.

**Comment by weiwei04** [comment]
```yaml
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: region
            operator: In
            values:
            - r0
```

The nodeAffinity term is `requiredDurinngSchedulingIgnoredDuringExecution` according to https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity

>  The "IgnoredDuringExecution" part of the names means that, similar to how nodeSelector works, if labels on a node change at runtime such that the affinity rules on a pod are no longer met, the pod will still continue to run on the node.

**Comment by adisky** [comment]
/triage accepted 
Needs to check how can we stop pod eviction if NodeAffinity not matched during kubelet startup

**Comment by pacoxu** [comment]
As @liggitt mentioned,  it may be fixed by #94087. 

You may try to reproduce it with #94087 (1.21.0+,  v1.20.3+, v1.19.8+,  v1.18.16+).

**Comment by rajaSahil** [comment]
@pacoxu Sure.

**Comment by rajaSahil** [comment]
@adisky @pacoxu 
Tested with k8s versions (1.21.0+, v1.20.3+, v1.19.8+, v1.18.16+).

#### Steps to reproduce:
1. label node0 with region=r0
2. schedule a pod A with nodeAffinity region=r0 to node0, and pod A up and running
3. label node0 with region=rx
4. restart kubelet on node0, after kubelet restart, pod A status becomes NodeAffinity.

Output:
```
$ kubectl get pods -owide                                                                                                                                                                                                   
NAME                 READY   STATUS         RESTARTS   AGE
test-node-affinity   0/1     NodeAffinity   0          63m

```

**Comment by pacoxu** [comment]
Oh, got it. Not the same issue as #94087.

I need to look into the code if it can be fixed reasonably.
